jupyterhub/apihandlers/users.py

```diff
@@ -710,19 +712,32 @@ async def get(self, user_name, server_name=''):
         # - spawner not running at all
         # - spawner failed
         # - spawner pending start (what we expect)
-        url = url_path_join(user.url, url_escape_path(server_name), '/')
-        ready_event = {
-            'progress': 100,
-            'ready': True,
-            'message': f"Server ready at {url}",
-            'html_message': 'Server ready at <a href="{0}">{0}</a>'.format(url),
-            'url': url,
-        }
         failed_event = {'progress': 100, 'failed': True, 'message': "Spawn failed"}
 
+        async def get_ready_event():
+            url = url_path_join(user.url, url_escape_path(server_name), '/')
+            ready_event = {
+                'progress': 100,
+                'ready': True,
+                'message': f"Server ready at {url}",
+                'html_message': 'Server ready at <a href="{0}">{0}</a>'.format(url),
+                'url': url,
+            }
+            original_ready_event = deepcopy(ready_event)
```

Comment:
deepcopy is fine here, but I don't think it is strictly required _in this case_ because all of the values are immutable scalars. So there isn't really a difference between a shallow and deep copy of this dict:

```python
In [8]: d = {"html_message": "Server ready"}

In [9]: d2 = d.copy()

In [10]: d2["html_message"] += " extra"

In [11]: d2["extra_key"] = "ok"

In [12]: d
Out[12]: {'html_message': 'Server ready'}

In [13]: d2
Out[13]: {'html_message': 'Server ready extra', 'extra_key': 'ok'}
```

deepcopy only makes a difference when the _values_ of the dict are themselves mutable, like lists, dicts, etc.